DomainParser/DomainParser/RulesParser.swift

```diff
@@ -9,41 +9,71 @@
 import Foundation
 
 
+struct ParsedRules {
+    /// Dictionary of rule arrays indexed by the last label of a rule.
+    let exceptions: [String: [Rule]]
+    /// Dictionary of rule arrays indexed by the last label of a rule.
+    let wildcardRules: [String: [Rule]]
+    /// Set of suffixes
+    let basicRules: Set<String>
+}
+
 class RulesParser {
-    
-    var exceptions = [Rule]()
-    var wildcardRules = [Rule]()
+
+    /// Dictionary of rule arrays indexed by the last label of a rule.
+    var exceptions = [String: [Rule]]()
+    /// Dictionary of rule arrays indexed by the last label of a rule.
+    var wildcardRules = [String: [Rule]]()
     /// Set of suffixes
     var basicRules = Set<String>()
     
     /// Parse the Data to extract an array of Rules. The array is sorted by importance.
-    func parse(raw: Data) throws -> ParsedRules {
+    func parse(raw: Data, sortRules: Bool) throws -> ParsedRules {
         guard let rulesText = String(data: raw, encoding: .utf8) else {
-            throw DomainParserError.parsingError(details: nil)
+            throw DomainParserError.ruleParsingError(message: "Can't parse rules data. Is it in UTF-8 format?")
         }
-        rulesText
-            .split(separator: "\n")
-            .forEach(parseRule)
-        return ParsedRules.init(exceptions: exceptions,
-                                wildcardRules: wildcardRules,
-                                basicRules: basicRules)
+
+        let allRules = rulesText.split(separator: "\n")
+        try allRules.forEach(parseRule)
+
+        if (sortRules) {
+            // Sort the collections from big to small so that the highest priority rules are first.
+            self.wildcardRules = self.wildcardRules.mapValues { (rules: [Rule]) in
```

Comment:
now that input data is sorted, is still necessary to sort it here?
I assume the rule will be inserted in sub dictionary in the same order as the file.

Reply:
Without this part of the code the test `testWildcardRulesSorting()` won't work.
But we are not sorting the rules again for the main `public_suffix_list.dat` file since we are calling it with `sortRules: false` in `DomainParser.swift`. The main rules are sorted just once in `UpdatePSL.swift`.

Reply:
I don't see the call of parse and sortRules in UpdatePSL. it seems you are doing the sorting there from the raw data, and only doing sort there which is better I think.

Reply:
got it, it can be confusing but it is ok.